AzureMonitor: Improve Log Analytics query efficiency

aangelisc opened the pull request:
This PR attempts to improve the efficiency of Log Analytics queries by ensuring that time-bounds exist for most queries.

If `Dashboard` is chosen as an option then a time-column must always be specified. The time-column will be selected by default based on the `timeSpan` value (if it exists), the first column with a `datetime` type, or `TimeGenerated` as the fallback. If the value doesn't exist or is invalid the query will succeed but may be less efficient.

- Removed the `Intersection` option and replaced with `Dashboard` to reduce confusion around how time-ranges work.
- Updated schema/types.
- Added migration to move `Intersection` option to `Dashboard` by default.
- Created `TimeManagement` component to hold this logic.
- Updated tests.

Additionally, I've simplified how the Kusto language worker is loaded by using async/await rather than callbacks. This *might* be easier to read. I've also improved the typings to make things clearer.

Fixes https://github.com/grafana/oss-plugin-partnerships/issues/190

![image](https://github.com/grafana/grafana/assets/15019026/13af78a2-f3d2-4fd8-b3b1-76d8df4c8a96)

# Release notice breaking change

This is a breaking change as we're removing support for `Intersection` (although it is replaced with an option that is nearly the same).

aangelisc committed:
Promisify loading schema

- Move schema loading to LogsQueryEditor
- Improve typing
- Switch callbacks to promises

aangelisc committed:
Update types

aangelisc committed:
Refactor backend for new props

- Rename intersectTime
- Support setting timeColumn
- Add additional properties to logs request body

aangelisc committed:
Update applyTemplateVariables

aangelisc committed:
Update set functions

aangelisc committed:
Add new TimeManagement component

aangelisc committed:
Update LogsQueryEditor

aangelisc committed:
Hardcode timestamp column for traces queries

aangelisc committed:
Ensure timeColumn is always set for log queries

aangelisc committed:
Update tests

aangelisc committed:
Update frontend tests

aangelisc committed:
Readd type to make migration easier

aangelisc committed:
Add migration

aangelisc committed:
Add fake schema

aangelisc committed:
Use predefined type

aangelisc committed:
Update checks and defaults

aangelisc committed:
Add tests

aangelisc committed:
README updates

aangelisc committed:
README update

aangelisc committed:
Type update

aangelisc committed:
Merge branch 'main' into andreas/az-logs-update

aangelisc committed:
Lint

aangelisc committed:
More linting and type fixing

aangelisc committed:
Error silently

aangelisc committed:
More linting and typing

aangelisc committed:
Update betterer

aangelisc committed:
Update test

aangelisc committed:
Simplify default column setting

aangelisc committed:
Fix default column setting

aangelisc committed:
Add tracking

asimpson commented on public/app/plugins/datasource/azuremonitor/dataquery.gen.ts:
```diff
@@ -162,7 +162,11 @@ export const defaultAzureMetricQuery: Partial<AzureMetricQuery> = {
+   */
+  dashboardTime?: boolean;
+  /**
+   * @deprecated Use dashboardTime instead
```
Do we need to mark this as deprecated or can we just remove it? We have a migration to remove it from existing queries, am I missing a reason to keep it around?

asimpson reviewed (approved):
:trophy: Looks good to me.

adamyeats commented on public/app/plugins/datasource/azuremonitor/components/LogsQueryEditor/TimeManagement.tsx:
```diff
@@ -0,0 +1,137 @@
+            }
+            return prev;
+          }, []);
+          timeColumns = timeColumns.concat(cols);
```
Mutating `timeColumns` that is returned from `useState` might be problematic, as these values are meant to be immutable. I'd recommend trying to refactor this to not mutate this value.

aangelisc replied:
This isn't mutating the state value, it's mutating the local variable. I'll rename for clarity.

aangelisc commented on public/app/plugins/datasource/azuremonitor/azure_log_analytics/azure_log_analytics_datasource.test.ts:
```diff
@@ -42,13 +48,13 @@ describe('AzureLogAnalyticsDatasource', () => {
     it('should return a schema to use with monaco-kusto', async () => {
       const result = await ctx.datasource.azureLogAnalyticsDatasource.getKustoSchema('myWorkspace');
```
Great catch! This will fix the type assertion. We still need the `?` as `database` can still be undefined but anything that removes a type assertion is a nice change 😊

aangelisc commented on public/app/plugins/datasource/azuremonitor/types/types.ts:
```diff
@@ -131,11 +134,32 @@ export interface AzureQueryEditorFieldProps {
   setError: (source: string, error: AzureMonitorErrorish | undefined) => void;
 }
 
+// To avoid a better issue we redeclare the EngineSchema type from @kusto/monaco-kusto
```
There's a typo there that should say `type` 😅 Will update the comment.

aangelisc commented on public/app/plugins/datasource/azuremonitor/components/LogsQueryEditor/QueryField.tsx:
```diff
@@ -17,50 +17,38 @@ interface MonacoLanguages {
+        const languages = monaco.languages as unknown as MonacoLanguages;
+        const model = editor.getModel();
+        if (model) {
+          const kustoWorker = await (await languages.kusto.getKustoWorker())(model?.uri);
```
Fair point 😊

aangelisc committed:
Review

- Fix typo on comment
- Destructure and remove type assertion
- Break out await into two variables
- Remove lets and rename variable for clarity

aangelisc committed:
Merge branch 'main' into andreas/az-logs-update
